client/mm/mm.go

```diff
@@ -88,34 +723,54 @@ func (m *MarketMaker) Run(ctx context.Context, cfgs []*BotConfig, pw []byte) err
 
 	m.ctx, m.die = context.WithCancel(ctx)
 
-	if err := duplicateBotConfig(cfgs); err != nil {
+	enabledCfgs, err := validateAndFilterEnabledConfigs(cfgs)
+	if err != nil {
 		return err
 	}
 
-	err := m.core.Login(pw)
+	if err := m.loginAndUnlockWallets(pw, enabledCfgs); err != nil {
+		return err
+	}
+
+	oracle, err := priceOracleFromConfigs(m.ctx, enabledCfgs, m.log.SubLogger("PriceOracle"))
 	if err != nil {
-		return fmt.Errorf("failed to login: %w", err)
+		return err
 	}
 
-	var oracle *priceOracle
-	marketsRequiringOracle := marketsRequiringPriceOracle(cfgs)
-	if len(marketsRequiringOracle) > 0 {
-		oracle, err = newPriceOracle(m.ctx, marketsRequiringOracle, m.log.SubLogger("PriceOracle"))
-		if err != nil {
-			return fmt.Errorf("failed to create PriceOracle: %v", err)
-		}
+	if err := m.setupBalances(enabledCfgs); err != nil {
+		return err
 	}
 
 	startedMarketMaking = true
 
 	wg := new(sync.WaitGroup)
-	for _, cfg := range cfgs {
+
+	// Listen for core notifications.
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		feed := m.core.NotificationFeed()
+		defer feed.ReturnFeed()
+
+		for {
+			select {
+			case <-m.ctx.Done():
+				return
+			case n := <-feed.C:
+				m.handleNotification(n)
+			}
+		}
+	}()
+
+	// Start each bot.
+	for _, cfg := range enabledCfgs {
 		switch {
 		case cfg.MMCfg != nil:
 			wg.Add(1)
 			go func(cfg *BotConfig) {
 				logger := m.log.SubLogger(fmt.Sprintf("MarketMaker-%s-%d-%d", cfg.Host, cfg.BaseAsset, cfg.QuoteAsset))
-				RunBasicMarketMaker(m.ctx, cfg, m.core, oracle, pw, logger)
+				mktID := dexMarketID(cfg.Host, cfg.BaseAsset, cfg.QuoteAsset)
+				RunBasicMarketMaker(m.ctx, cfg, m.wrappedCoreForBot(mktID), oracle, pw, logger)
 				wg.Done()
 			}(cfg)
```

Comment:
Is there any way to update the bot config, or do you have to shut down and restart?

Reply:
Currently you have to shut down and restart.